### PR TITLE
Extended prediction algorithm in Initial Resource to use also namespace

### PR DESCRIPTION
**Pull request by piosz**
part of #13589

@mwielgus could you please take a look? In the mean time I'll test the change manually.

cc @burmanm

### CODE REVIEW

File: plugin/pkg/admission/initialresources/admission.go [hunk 1]
```diff
@@ -32,8 +32,9 @@ import (
 )
 
 var (
-	source     = flag.String("ir-data-source", "influxdb", "Data source used by InitialResources. Supported options: influxdb.")
+	source     = flag.String("ir-data-source", "influxdb", "Data source used by InitialResources. Supported options: influxdb, gcm.")
 	percentile = flag.Int64("ir-percentile", 90, "Which percentile of samples should InitialResources use when estimating resources. For experiment purposes.")
+	nsAware    = flag.Bool("ir-namespace-aware", false, "Whether the estimation should be made only based on data from the same namespace.")
```
Comment (mwielgus) [inline plugin/pkg/admission/initialresources/admission.go]: ir-namespace-only

File: plugin/pkg/admission/initialresources/admission.go [hunk 2]
```diff
@@ -138,29 +143,47 @@ func (ir initialResources) estimateAndFillResourcesIfNotSet(pod *api.Pod) {
...
+			}
+		}
+		if samples < samplesThreshold {
+			// Historical data from last 30 days for the same image:tag within all namespaces.
+			start := end.Add(-month)
+			if usage, samples, err = ir.source.GetUsagePercentile(kind, ir.percentile, c.Image, "", true, start, end); err != nil {
+				return nil, err
+			}
+		}
+		if samples < samplesThreshold {
```
Comment (mwielgus) [inline plugin/pkg/admission/initialresources/admission.go]: there should also be same image (no tag) within the specific namespace, if ir.nsAware/ns.Only

File: plugin/pkg/admission/initialresources/admission.go (hunk 1 above)
Comment (piosz) [inline plugin/pkg/admission/initialresources/admission.go]: Done

File: plugin/pkg/admission/initialresources/admission.go (hunk 2 above)
Comment (piosz) [inline plugin/pkg/admission/initialresources/admission.go]: Done

**Comment by piosz** [comment]
Tested manually both on InfluxDB and GCM and the plugin works.

File: plugin/pkg/admission/initialresources/admission.go [hunk 3]
```diff
@@ -138,29 +143,56 @@ func (ir initialResources) estimateAndFillResourcesIfNotSet(pod *api.Pod) {
...
+			// Historical data from last 30 days for the same image within all namespaces.
+			start := end.Add(-month)
+			image := strings.Split(c.Image, ":")[0]
+			if usage, samples, err = ir.source.GetUsagePercentile(kind, ir.percentile, image, "", false, start, end); err != nil {
+				return nil, err
+			}
+		}
+	} else {
```
Comment (mwielgus) [inline plugin/pkg/admission/initialresources/admission.go]: nit: should be if ir.nsOnly { .... } else {...}. Currently it is if !ir.nsOnly {....}.

**Comment by mwielgus** [comment]
LGTM

File: plugin/pkg/admission/initialresources/admission.go (hunk 3 above)
Comment (piosz) [inline plugin/pkg/admission/initialresources/admission.go]: Done

**Comment by burmanm** [comment]
Yeps, looks good to me as well.

**Pull request merged**